core/json/mod.rs

```diff
@@ -186,8 +186,57 @@ pub fn json_extract(value: &OwnedValue, paths: &[OwnedValue]) -> crate::Result<O
     Ok(OwnedValue::Text(LimboText::json(Rc::new(result))))
 }
 
-fn json_extract_single(json: &Val, path: &str) -> crate::Result<Val> {
-    let json_path = json_path(path)?;
+pub fn json_type(value: &OwnedValue, path: Option<&OwnedValue>) -> crate::Result<OwnedValue> {
+    if let OwnedValue::Null = value {
+        return Ok(OwnedValue::Null);
+    }
+
+    let json = get_json_value(value)?;
+
+    let json = if let Some(path) = path {
+        match json_extract_single(&json, path)? {
+            ExtractResult::Val(val) => val,
+            ExtractResult::None => return Ok(OwnedValue::Null),
+        }
+    } else {
+        &json
+    };
+
+    let val = match json {
+        Val::Null => "null",
+        Val::Bool(v) => {
+            if *v {
+                "true"
+            } else {
+                "false"
+            }
+        }
+        Val::Integer(_) => "integer",
+        Val::Float(_) => "real",
+        Val::String(_) => "text",
+        Val::Array(_) => "array",
+        Val::Object(_) => "object",
+    };
+
+    Ok(OwnedValue::Text(LimboText::json(Rc::new(val.to_string()))))
+}
+
+/// Represents the result of a JSON extraction
+#[derive(Clone, Debug)]
+enum ExtractResult<'a> {
+    /// Val represents a successful extraction of a JSON value
+    Val(&'a Val),
+    /// None means that the given path did not exist in the JSON, opposed to a path that exists
+    /// and points to a null value
+    None,
+}
+
+fn json_extract_single<'a>(json: &'a Val, path: &OwnedValue) -> crate::Result<ExtractResult<'a>> {
```

Comment:
Before this PR, `json_extract_single` returned a `crate::Result<Val>`. 

I added lifetime specifiers to ExtractResult so that we don't need an allocation here, but Clippy suggests that the compiler will derefence the result anyway:

```
Ok(ExtractResult::Val(&current_element))

This expression creates a reference which is immediately dereferenced by the compiler Help: for further information visit https:// rust-lang. github. io/ rust-clippy/ master/ index. html#needless_borrow
```

Any idea what to do here? 🤔

Reply:
Oh wait I can also just use `Option` instead of creating my own enum 🤦